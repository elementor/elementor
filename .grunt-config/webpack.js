/**
 * Grunt webpack task config
 * @package Elementor
 */
const path = require( 'path' );

const packagesConfigs = require( './webpack.packages.js' );

// Handling minification for production assets.
const TerserPlugin = require( 'terser-webpack-plugin' );

const aliasList = require('./webpack.alias.js').resolve;

const webpack = require('webpack');

// Cleaning up existing chunks before creating new ones.
const RemoveChunksPlugin = require('./remove-chunks');

const WatchTimePlugin = require('./plugins/watch-time/index');

// Preventing auto-generated long names of shared sub chunks (optimization.splitChunks.minChunks) by using only the hash.
const getChunkName = ( chunkData, environment ) => {
	const minSuffix = 'production' === environment ? '.min' : '',
		name = chunkData.chunk.name ? '[name].' : '';

	return `${ name }[contenthash].bundle${ minSuffix }.js`;
};

const getModuleRules = ( presets ) => {
	return {
		rules: [
			{
				test: /app.*\.(s)?css$/i,
				use: [
					{
						loader: './loaders/app-imports.js',
					},
				],
			},
			{
				test: /\.js$/,
				exclude: /node_modules/,
				use: [
					{
						loader: 'babel-loader',
						options: {
							presets,
							plugins: [
								[ '@wordpress/babel-plugin-import-jsx-pragma' ],
								[ '@babel/plugin-transform-react-jsx', {
									'pragmaFrag': 'React.Fragment',
								} ],
								[ '@babel/plugin-transform-runtime' ],
								[ '@babel/plugin-transform-modules-commonjs' ],
							],
						},
					},
				],
			},
		],
	};
};

const moduleRules = getModuleRules( [ '@wordpress/default' ] );

const frontendRulesPresets = [ [
	'@babel/preset-env',
	{
		targets: {
			browsers: [
				'last 1 Android versions',
				'last 1 ChromeAndroid versions',
				'last 2 Chrome versions',
				'last 2 Firefox versions',
				'Safari >= 14',
				'iOS >= 14',
				'last 2 Edge versions',
				'last 2 Opera versions',
			],
		},
		"useBuiltIns": "usage",
		"corejs": "3.23",
	}
] ];

const frontendModuleRules = getModuleRules( frontendRulesPresets );

const entry = {
	'editor': [
		path.resolve( __dirname, '../assets/dev/js/editor/utils/jquery-serialize-object.js' ),
		path.resolve( __dirname, '../assets/dev/js/editor/utils/jquery-html5-dnd.js' ),
		path.resolve( __dirname, '../assets/dev/js/editor/editor.js' ),
	],
	'admin': path.resolve( __dirname, '../assets/dev/js/admin/admin.js' ),
	'admin-feedback': path.resolve( __dirname, '../assets/dev/js/admin/admin-feedback.js' ),
	'announcements-app': path.resolve( __dirname, '../modules/announcements/assets/js/index.js' ),
	'common': path.resolve( __dirname, '../core/common/assets/js/common.js' ),
	'dev-tools': path.resolve( __dirname, '../modules/dev-tools/assets/js/index.js' ),
	'elementor-admin-bar': path.resolve( __dirname, '../modules/admin-bar/assets/js/frontend/module.js' ),
	'gutenberg': path.resolve( __dirname, '../assets/dev/js/admin/gutenberg.js' ),
	'new-template': path.resolve( __dirname, '../assets/dev/js/admin/new-template/new-template.js' ),
	'app': path.resolve( __dirname, '../app/assets/js/index.js' ),
	'app-loader': path.resolve( __dirname, '../app/assets/js/app-loader' ),
	'app-packages': path.resolve( __dirname, '../app/assets/js/app-packages' ),
	'beta-tester': path.resolve( __dirname, '../assets/dev/js/admin/beta-tester/beta-tester.js' ),
	'common-modules': path.resolve( __dirname, '../core/common/assets/js/modules' ),
	'editor-modules': path.resolve( __dirname, '../assets/dev/js/editor/modules.js' ),
	'admin-modules': path.resolve( __dirname, '../assets/dev/js/admin/modules.js' ),
	'editor-document': path.resolve( __dirname, '../assets/dev/js/editor/editor-document.js' ),
	'qunit-tests': path.resolve( __dirname, '../tests/qunit/main.js' ),
	'admin-top-bar': path.resolve( __dirname, '../modules/admin-top-bar/assets/js/admin.js' ),
	'checklist': path.resolve( __dirname, '../modules/checklist/assets/js/editor.js' ),
	'nested-elements': path.resolve( __dirname, '../modules/nested-elements/assets/js/editor/index.js' ),
	'nested-tabs': path.resolve( __dirname, '../modules/nested-tabs/assets/js/editor/index.js' ),
	'nested-accordion': path.resolve( __dirname, '../modules/nested-accordion/assets/js/editor/index.js' ),
	'container-converter': path.resolve( __dirname, '../modules/container-converter/assets/js/editor/module.js' ),
	'atomic-widgets-editor': path.resolve( __dirname, '../modules/atomic-widgets/assets/js/editor/module.js' ),
	'notes': path.resolve( __dirname, '../modules/notes/assets/js/notes.js' ),
	'web-cli': path.resolve( __dirname, '../modules/web-cli/assets/js/index.js' ),
	'import-export-admin': path.resolve( __dirname, '../app/modules/import-export/assets/js/admin.js' ),
	'kit-elements-defaults-editor': path.resolve( __dirname, '../modules/kit-elements-defaults/assets/js/editor/index.js' ),
	'editor-loader-v1': path.resolve( __dirname, '../core/editor/loader/v1/js/editor-loader-v1.js' ),
	'editor-loader-v2': path.resolve( __dirname, '../core/editor/loader/v2/js/editor-loader-v2.js' ),
	'editor-environment-v2': path.resolve( __dirname, '../core/editor/loader/v2/js/editor-environment-v2.js' ),
	'responsive-bar': path.resolve( __dirname, '../assets/dev/js/editor/regions/responsive-bar/index.js' ),
	'ai': path.resolve( __dirname, '../modules/ai/assets/js/editor/index.js' ),
	'admin-notifications': path.resolve( __dirname, '../modules/notifications/assets/js/admin.js' ),
	'editor-notifications': path.resolve( __dirname, '../modules/notifications/assets/js/editor.js' ),
	'ai-layout': path.resolve( __dirname, '../modules/ai/assets/js/editor/layout-module.js' ),
	'ai-gutenberg': path.resolve( __dirname, '../modules/ai/assets/js/gutenberg/index.js' ),
	'element-manager-admin': path.resolve( __dirname, '../modules/element-manager/assets/js/admin.js' ),
	'media-hints': path.resolve( __dirname, '../assets/dev/js/admin/hints/media.js' ),
	'ai-media-library': path.resolve( __dirname, '../modules/ai/assets/js/media-library/index.js' ),
	// Temporary solution for the AI App in the Admin.
	'ai-admin': path.resolve( __dirname, '../modules/ai/assets/js/admin/index.js' ),
	'styleguide': path.resolve( __dirname, '../modules/styleguide/assets/js/styleguide.js' ),
	'styleguide-app-initiator': path.resolve( __dirname, '../modules/styleguide/assets/js/styleguide-app-initiator.js' ),
	'e-home-screen': path.resolve( __dirname, '../modules/home/assets/js/app.js' ),
	'e-wc-product-editor': path.resolve( __dirname, '../modules/wc-product-editor/assets/js/e-wc-product-editor.js' ),
	'floating-elements-modal': path.resolve( __dirname, '../assets/dev/js/admin/floating-elements/new-floating-elements.js' ),
};

const frontendEntries = {
	'frontend-modules': path.resolve( __dirname, '../assets/dev/js/frontend/modules.js' ),
	'frontend': { import: path.resolve( __dirname, '../assets/dev/js/frontend/frontend.js' ), dependOn: 'frontend-modules' },
};

const externals = [
	{
		'@wordpress/i18n': 'wp.i18n',
		react: 'React',
		'react-dom': 'ReactDOM',
		'@elementor/app-ui': 'elementorAppPackages.appUi',
		'@elementor/components': 'elementorAppPackages.components',
		'@elementor/hooks': 'elementorAppPackages.hooks',
		'@elementor/site-editor': 'elementorAppPackages.siteEditor',
		'@elementor/router': 'elementorAppPackages.router',
		'@elementor/ui': 'elementorV2.ui',
		'@elementor/icons': 'elementorV2.icons',
		'@elementor/editor-app-bar': 'elementorV2.editorAppBar',
		'@elementor/editor-v1-adapters': 'elementorV2.editorV1Adapters',
		'@wordpress/dom-ready': 'wp.domReady',
		'@wordpress/components': 'wp.components',
		'@wordpress/core-data': 'wp.coreData',
		'@wordpress/data': 'wp.data',
		'@wordpress/plugins': 'wp.plugins',
		'@woocommerce/admin-layout': 'wc.adminLayout',
	},
	// Handle tree-shaking imports for ui and icons packages (@elementor/ui/xxx) to be pointed to the external object (elementorV2.ui.xxx).
	function ( { request }, callback ) {
		const matches = request.match( /^@elementor\/(ui|icons)\/(.+)$/ );

		if ( matches?.length ) {
			return callback( null, `elementorV2.${ matches[ 1 ] }['${ matches[ 2 ] }']` );
		}

		callback();
	},
];

const plugins = [
	new webpack.ProvidePlugin( {
		React: 'react',
		ReactDOM: 'react-dom',
		PropTypes: 'prop-types',
		__: ['@wordpress/i18n', '__'],
		sprintf: ['@wordpress/i18n', 'sprintf'],
	} ),
	new WatchTimePlugin(),
];

const baseConfig = {
	target: 'web',
	context: __dirname,
	externals,
	resolve: aliasList,
};

const devSharedConfig = {
	...baseConfig,
	devtool: 'source-map',
	mode: 'development',
	output: {
		path: path.resolve( __dirname, '../assets/js' ),
		chunkFilename: ( chunkData ) => getChunkName( chunkData, 'development' ),
		filename: '[name].js',
		devtoolModuleFilenameTemplate: '../[resource]',
		// Prevents the collision of chunk names between different bundles.
		uniqueName: 'elementor',
	},
	watch: true,
};

const webpackConfig = [
	{
		...devSharedConfig,
		module: moduleRules,
		plugins: [
			...plugins,
		],
		name: 'base',
		entry: entry,
	},
	{
		...devSharedConfig,
		module: frontendModuleRules,
		plugins: [
			new RemoveChunksPlugin( '.bundle.js' ),
			...plugins,
		],
		name: 'frontend',
		optimization: {
			runtimeChunk:  {
				name: 'webpack.runtime',
			},
			splitChunks: {
				minChunks: 2,
			},
		},
		entry: frontendEntries,
	},
	packagesConfigs.dev,
];

const prodSharedOptimization = {
	minimize: true,
	minimizer: [
		new TerserPlugin( {
			terserOptions: {
				keep_fnames: true,
			},
			include: /\.min\.js$/
		} ),
	],
};

const prodSharedConfig = {
	...baseConfig,
	mode: 'production',
	output: {
		path: path.resolve( __dirname, '../assets/js' ),
		chunkFilename: ( chunkData ) => getChunkName( chunkData, 'production' ),
		filename: '[name].js',
		// Prevents the collision of chunk names between different bundles.
		uniqueName: 'elementor',
	},
	performance: { hints: false },
};

const webpackProductionConfig = [
	{
		...prodSharedConfig,
		module: moduleRules,
		plugins: [
			...plugins,
		],
		name: 'base',
		entry: {
			// Clone.
			...entry,
		},
		optimization: {
			...prodSharedOptimization,
		},
	},
	{
		...prodSharedConfig,
		module: frontendModuleRules,
		plugins: [
			new RemoveChunksPlugin( '.bundle.min.js' ),
			...plugins,
		],
		name: 'frontend',
		entry: {
			// Clone.
			...frontendEntries,
		},
		optimization: {
			...prodSharedOptimization,
			runtimeChunk: {
				name: 'webpack.runtime.min',
			},
			splitChunks: {
				minChunks: 2,
			},
		},
	},
];

// Adding .min suffix to production entries.
webpackProductionConfig.forEach( ( config, index ) => {
	for ( const entryPoint in config.entry ) {
		let entryValue = config.entry[ entryPoint ];

		if ( entryValue.dependOn ) {
			// We duplicate the 'entryValue' obj for not affecting the 'entry' obj used by the dev process.
			entryValue = { ...entryValue };

			entryValue.dependOn += '.min';
		}

		delete config.entry[ entryPoint ];
		config.entry[ entryPoint + '.min' ] = entryValue;
	}
} );

// The 'packages' config doesn't need a `.min` suffix (it has its own config).
webpackProductionConfig.push( packagesConfigs.prod );

const developmentNoWatchConfig = webpackConfig.map( ( config ) => {
	return { ...config, watch: false };
} );

const productionWatchConfig = webpackProductionConfig.map( ( config ) => {
	return { ...config, watch: true };
} );

const gruntWebpackConfig = {
	development: webpackConfig,
	developmentNoWatch: developmentNoWatchConfig,
	production: webpackProductionConfig,
	productionWatch: productionWatchConfig,
	packages: packagesConfigs.dev,
};

module.exports = gruntWebpackConfig;
