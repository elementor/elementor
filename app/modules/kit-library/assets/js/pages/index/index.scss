$e-kit-library-sticky-search-z-index: 2;

.e-kit-library {
	&__index-layout-container {
		flex-grow: 1;
		overflow-y: auto;
		display: flex;
		flex-direction: column;
	}

	&__index-layout-top-area {
		padding: spacing(30) spacing(44);
		position: sticky;
		top: -1px;
		width: 100%;
		z-index: $e-kit-library-sticky-search-z-index;
		background-color: var(--app-background-color);
		gap: spacing( 24 );

		&-search, &-sort {
			min-width: 265px;
		}

		&-search {
			flex: 1;
		}
	}

	&__index-layout-main {
		padding-top: spacing(0);
		padding-bottom: spacing(24);
		overflow-y: hidden;
		height: auto;
		flex: 1 0 auto;
		display: flex;
		flex-direction: column;
		justify-content: space-between;
	}
}
