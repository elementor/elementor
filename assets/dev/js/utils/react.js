import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { createRoot } from 'react-dom/client';

/**
 * Support conditional rendering of a React App to the DOM, based on the React version.
 * We use `createRoot` when available, but fallback to `ReactDOM.render` for older versions.
 *
 * @param { React.ReactElement } app        The app to render.
 * @param { HTMLElement }        domElement The DOM element to render the app into.
 *
 * @return {{ unmount: () => void }} The unmount function.
 */
function render( app, domElement ) {
	let unmountFunction;

	try {
		const root = createRoot( domElement );
		root.render( app );

		unmountFunction = () => {
			root.unmount();
		};
	} catch ( e ) {
		// eslint-disable-next-line react/no-deprecated
		ReactDOM.render( app, domElement );

		unmountFunction = () => {
			// eslint-disable-next-line react/no-deprecated
			ReactDOM.unmountComponentAtNode( domElement );
		};
	}

	return { unmount: unmountFunction };
}

export default {
	render,
};
