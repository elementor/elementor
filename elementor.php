<?php
/**
 * Plugin Name: Elementor
 * Description: The Elementor Website Builder has it all: drag and drop page builder, pixel perfect design, mobile responsive editing, and more. Get started now!
 * Plugin URI: https://elementor.com/?utm_source=wp-plugins&utm_campaign=plugin-uri&utm_medium=wp-dash
 * Author: Elementor.com
 * Version: 3.12.1
 * Author URI: https://elementor.com/?utm_source=wp-plugins&utm_campaign=author-uri&utm_medium=wp-dash
 *
 * Text Domain: elementor
 *
 * @package Elementor
 * @category Core
 *
 * Elementor is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * Elementor is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}

define( 'ELEMENTOR_VERSION', '3.12.1' );

define( 'ELEMENTOR__FILE__', __FILE__ );
define( 'ELEMENTOR_PLUGIN_BASE', plugin_basename( ELEMENTOR__FILE__ ) );
define( 'ELEMENTOR_PATH', plugin_dir_path( ELEMENTOR__FILE__ ) );

if ( defined( 'ELEMENTOR_TESTS' ) && ELEMENTOR_TESTS ) {
	define( 'ELEMENTOR_URL', 'file://' . ELEMENTOR_PATH );
} else {
	define( 'ELEMENTOR_URL', plugins_url( '/', ELEMENTOR__FILE__ ) );
}

define( 'ELEMENTOR_MODULES_PATH', plugin_dir_path( ELEMENTOR__FILE__ ) . '/modules' );
define( 'ELEMENTOR_ASSETS_PATH', ELEMENTOR_PATH . 'assets/' );
define( 'ELEMENTOR_ASSETS_URL', ELEMENTOR_URL . 'assets/' );

add_action( 'plugins_loaded', 'elementor_load_plugin_textdomain' );

if ( ! version_compare( PHP_VERSION, '7.0', '>=' ) ) {
	add_action( 'admin_notices', 'elementor_fail_php_version' );
} elseif ( ! version_compare( get_bloginfo( 'version' ), '5.9', '>=' ) ) {
	add_action( 'admin_notices', 'elementor_fail_wp_version' );
} else {
	require ELEMENTOR_PATH . 'includes/plugin.php';
}

/**
 * Load Elementor textdomain.
 *
 * Load gettext translate for Elementor text domain.
 *
 * @since 1.0.0
 *
 * @return void
 */
function elementor_load_plugin_textdomain() {
	load_plugin_textdomain( 'elementor' );
}

/**
 * Elementor admin notice for minimum PHP version.
 *
 * Warning when the site doesn't have the minimum required PHP version.
 *
 * @since 1.0.0
 *
 * @return void
 */
function elementor_fail_php_version() {
	$message = sprintf(
		/* translators: 1: `<h3>` opening tag, 2: `</h3>` closing tag, 3: PHP version. 4: Link opening tag, 5: Link closing tag. */
		esc_html__( '%1$sElementor isn’t running because PHP is outdated.%2$s Update to PHP version %3$s and get back to creating! %4$sShow me how%5$s', 'elementor' ),
		'<h3>',
		'</h3>',
		'7.0',
		'<a href="https://go.elementor.com/wp-dash-update-php/" target="_blank">',
		'</a>'
	);
	$html_message = sprintf( '<div class="error">%s</div>', wpautop( $message ) );
	echo wp_kses_post( $html_message );
}

/**
 * Elementor admin notice for minimum WordPress version.
 *
 * Warning when the site doesn't have the minimum required WordPress version.
 *
 * @since 1.5.0
 *
 * @return void
 */
function elementor_fail_wp_version() {
	$message = sprintf(
		/* translators: 1: `<h3>` opening tag, 2: `</h3>` closing tag, 3: WP version. 4: Link opening tag, 5: Link closing tag. */
		esc_html__( '%1$sElementor isn’t running because WordPress is outdated.%2$s Update to version %3$s and get back to creating! %4$sShow me how%5$s', 'elementor' ),
		'<h3>',
		'</h3>',
		'5.9',
		'<a href="https://go.elementor.com/wp-dash-update-wordpress/" target="_blank">',
		'</a>'
	);
	$html_message = sprintf( '<div class="error">%s</div>', wpautop( $message ) );
	echo wp_kses_post( $html_message );
}
