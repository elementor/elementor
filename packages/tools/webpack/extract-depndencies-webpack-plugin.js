// Inspired by "Dependency Extraction Webpack Plugin" by @wordpress team.
// Link: https://github.com/WordPress/gutenberg/tree/trunk/packages/dependency-extraction-webpack-plugin

const { sources: { RawSource } } = require( 'webpack' );

module.exports = class ExtractDependenciesWebpackPlugin {
	generateHandleName;
	replaceDependencyNames;
	generateAssetsFileName;

	constructor( {
		generateHandleName,
		replaceDependencyNames,
		generateAssetsFileName,
	} = {} ) {
		this.generateHandleName = generateHandleName || this.defaultGenerateHandleName;
		this.replaceDependencyNames = replaceDependencyNames || this.defaultReplaceDependencyNames;
		this.generateAssetsFileName = generateAssetsFileName || this.defaultGenerateAssetsFileName;
	}

	apply( compiler ) {
		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
			compilation.hooks.processAssets.tap( { name: this.constructor.name }, () => {
				[ ...compilation.entrypoints ].forEach( ( [ id, entrypoint ] ) => {
					const chunk = entrypoint.chunks.find( ( { name } ) => name === id );
					const chunkJSFile = this.getFileFromChunk( chunk );

					if ( ! chunkJSFile ) {
						return;
					}

					const deps = this.getDepsFromChunk( compilation, chunk );

					const assetFilename = this.generateAssetsFileName(
						compilation.getPath( '[file]', { filename: chunkJSFile } )
					);

					const content = this.createAssetsFileContent(
						id,
						deps,
						compilation.options.optimization.minimize
					);

					// Add source and file into compilation for webpack to output.
					compilation.assets[ assetFilename ] = new RawSource( content );

					chunk.files.add( assetFilename );
				} );
			} );
		} );
	}

	getDepsFromChunk( compilation, chunk ) {
		const externals = Object.keys( compilation.options.externals );
		const depsSet = new Set();

		compilation.chunkGraph.getChunkModules( chunk ).forEach( ( module ) => {
			[ ...( module.modules || [] ), module ].forEach( ( subModule ) => {
				const isExternalDep = externals.includes( subModule.userRequest );

				if ( ! isExternalDep ) {
					return;
				}

				depsSet.add( subModule.userRequest );
			} );
		} );

		return depsSet;
	}

	createAssetsFileContent( entryId, deps, shouldMinify ) {
		const handleName = this.generateHandleName( entryId );

		const depsAsString = [ ...deps ]
			.map( ( dep ) => this.replaceDependencyNames( dep ) )
			.filter( ( dep ) => dep !== handleName )
			.sort()
			.map( ( dep ) => `'${ dep }',` )
			.join( '\n\t\t' );

		let content =
`<?php
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

return [
	'handle' => '${ handleName }',
	'deps' => [
		${ depsAsString }
	],
];
`;

		if ( shouldMinify ) {
			content = content
				.replaceAll( /\s/gm, '' )
				.replace( '<?php', '<?php\n' );
		}

		return content;
	}

	getFileFromChunk( chunk ) {
		return [ ...chunk.files ].find( ( f ) => /\.js$/i.test( f ) );
	}

	defaultReplaceDependencyNames( name ) {
		const map = new Map( [
			[ '@elementor/', 'elementor-packages-' ],
			[ '@wordpress/', 'wp-' ],
		] );

		for ( const [ key, value ] of map ) {
			if ( name.startsWith( key ) ) {
				return name.replace( key, value );
			}
		}

		return name;
	}

	defaultGenerateAssetsFileName( filename ) {
		return filename.replace( /(\.min)?\.js$/i, '.asset.php' );
	}

	defaultGenerateHandleName( name ) {
		return `elementor-packages-${ name }`;
	}
};
