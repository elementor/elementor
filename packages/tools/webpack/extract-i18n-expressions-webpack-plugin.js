const fs = require( 'fs' );
const path = require( 'path' );

const MODULE_FILTERS = Object.freeze( [ /(([^!?\s]+?)(?:\.js|\.jsx|\.ts|\.tsx))$/, /^((?!node_modules).)*$/ ] );

const COMMENTS_REGEXPS = Object.freeze( [
	// Matches translators comment block: `/* translators: %s */`.
	/\/\*[\t ]*translators:.*\*\//gm,
	/// Matches translators inline comment: `// translators: %s`.
	/(\/\/)[\t ]*translators:[^\r\n]*/gm,
] );

const TRANSLATIONS_REGEXPS = Object.freeze( [
	// Matches translation functions: `__('Hello', 'elementor')`, `_n('Me', 'Us', 2, 'elementor-pro')`.
	/\b_(?:_|n|nx|x)\(.*?,\s*(?<c>['"`])[\w-]+\k<c>\)/gm,
] );

module.exports = class ExtractI18nExpressionsWebpackPlugin {
	apply( compiler ) {
		// Learn more about Webpack plugin system: https://webpack.js.org/api/plugins/

		let translationCallExpressions = [];

		// Learn more about Webpack compilation process and hooks: https://webpack.js.org/api/compilation-hooks/
		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
			// We tap into the time that Webpack has finished processing all the other assets
			// learn more: https://webpack.js.org/api/compilation-hooks/#processassets.
			compilation.hooks.processAssets.tap( { name: this.constructor.name }, () => {
				translationCallExpressions = this.getModuleExpressionsMap( compilation );
			} );
		} );

		compiler.hooks.afterEmit.tapPromise( this.constructor.name, async ( compilation ) => {
			// Create all the translations files based on the call expressions.
			await this.createTranslationsFiles( compilation, translationCallExpressions );
		} );
	}

	getModuleExpressionsMap( compilation ) {
		const moduleExpressionsMap = new Map();

		[ ...compilation.chunks ].forEach( ( chunk ) => {
			const chunkJSFile = this.getFileFromChunk( chunk );

			if ( ! chunkJSFile ) {
				// There's no JS file in this chunk, no work for us. Typically a `style.css` from cache group.
				return;
			}

			compilation.chunkGraph.getChunkModules( chunk ).forEach( ( module ) => {
				this.getSubModulesToCheck( module ).forEach( ( subModule ) => {
					const mainEntryFile = this.findMainModuleOfEntry( subModule, compilation );

					if ( ! moduleExpressionsMap.has( mainEntryFile ) ) {
						moduleExpressionsMap.set( mainEntryFile, [] );
					}

					// Running over the submodules and find all the translation call expressions and their translators comment
					// (e.g `/* translators: %s: name*/ __('Hello %s', 'elementor')`),
					// extract them and add them to a Map, where the key is the main entry file, and the value is an array of all the
					// translation call expressions.
					this.extractExpressionsFromSubmodule( subModule ).forEach( ( expression ) => {
						moduleExpressionsMap.get( mainEntryFile ).push( expression );
					} );
				} );
			} );
		} );

		return moduleExpressionsMap;
	}

	async createTranslationsFiles( compilation, translationCallExpressions ) {
		const promises = [ ...compilation.entrypoints ].map( ( [ id, entrypoint ] ) => {
			const chunk = entrypoint.chunks.find( ( { name } ) => name === id );

			const chunkJSFile = this.getFileFromChunk( chunk );

			if ( ! chunkJSFile ) {
				return Promise.resolve();
			}

			const mainFilePath = compilation.options.entry[ id ].import[ 0 ];

			const assetFilename = this.generateTranslationFilename(
				compilation.options.output.path,
				compilation.getPath( '[file]', { filename: chunkJSFile } )
			);

			const assetFileContent = this.generateTranslationFileContent(
				translationCallExpressions.get( mainFilePath ) || []
			);

			return fs.promises.writeFile( assetFilename, assetFileContent );
		} );

		return await Promise.all( promises );
	}

	getSubModulesToCheck( module ) {
		return [ ...( module.modules || [] ), module ]
			.filter( ( subModule ) => this.shouldCheckModule( subModule ) );
	}

	getFileFromChunk( chunk ) {
		return [ ...chunk.files ].find( ( f ) => /\.js$/i.test( f ) );
	}

	shouldCheckModule( module ) {
		return MODULE_FILTERS.every( ( filter ) => !! filter.test( module.userRequest ) );
	}

	findMainModuleOfEntry( module, compilation ) {
		if ( compilation.moduleGraph.getIssuer( module ) ) {
			return this.findMainModuleOfEntry( compilation.moduleGraph.getIssuer( module ), compilation );
		}

		return module.rawRequest;
	}

	extractExpressionsFromSubmodule( subModule ) {
		const source = subModule?._source?._valueAsString;

		if ( ! source ) {
			return [];
		}

		const translationCallExpressions = [];

		[
			...TRANSLATIONS_REGEXPS,
			...COMMENTS_REGEXPS,
		].forEach( ( regexp ) => {
			[ ...source.matchAll( regexp ) ].forEach( ( res ) => {
				translationCallExpressions.push( {
					type: COMMENTS_REGEXPS.includes( regexp ) ? 'comment' : 'call-expression',
					index: res.index,
					value: res[ 0 ],
				} );
			} );
		} );

		return translationCallExpressions;
	}

	generateTranslationFilename( basePath, filename ) {
		return path.join(
			basePath,
			filename.replace( /(\.min)?\.js$/i, '.strings.js' )
		);
	}

	generateTranslationFileContent( expressions ) {
		return expressions
			// Sort by the index it was found in the file based on the regexp (and not by the order it was added to the array).
			.sort( ( a, b ) => a.index - b.index )
			// Add a semicolon when needed.
			.map( ( expr ) => `${ expr.value }${ expr.type === 'comment' ? '' : ';' }` )
			// Join all the expressions to a single string with line-breaks between them.
			.join( '\n' );
	}
};
