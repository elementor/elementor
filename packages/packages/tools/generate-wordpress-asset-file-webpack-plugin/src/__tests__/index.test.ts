/**
 * @jest-environment node
 */

import * as fs from 'fs';
import { vol } from 'memfs';
import * as path from 'path';
import { webpack } from 'webpack';

import { GenerateWordPressAssetFileWebpackPlugin } from '../index';

jest.mock( 'fs', () => jest.requireActual( 'memfs' ) );

describe( '@elementor/generate-wordpress-asset-file-webpack-plugin', () => {
	beforeEach( () => {
		const fileContent = `
			import elementor from '@elementor/editor';
			import wp from '@wordpress/element';
			import other from '@other/package';
			import react from 'react';
			import ReactDOM from 'react-dom';

			elementor();
			wp();
			other();
		`;

		fs.writeFileSync( path.resolve( '/app.js' ), fileContent );
		fs.writeFileSync( path.resolve( '/extension.js' ), fileContent );
		fs.writeFileSync( path.resolve( '/util.js' ), fileContent );
	} );

	afterEach( () => {
		vol.reset();
	} );

	it( 'should generate assets files', ( done ) => {
		// Arrange.
		const compiler = webpack( {
			mode: 'development',
			entry: {
				app: path.resolve( '/app.js' ),
				extension: path.resolve( '/extension.js' ),
				util: path.resolve( '/util.js' ),
			},
			output: {
				filename: '[name].js',
				path: path.resolve( '/dist' ),
			},
			externals: {
				// Required so webpack won't try to resolve those packages, which don't exist.
				'@elementor/editor': 'editor',
				'@wordpress/element': 'wp',
				'@other/package': 'other',
				react: 'react',
				'react-dom': 'reactDOM',
			},
			plugins: [
				new GenerateWordPressAssetFileWebpackPlugin( {
					handle: ( entryName ) => `elementor-v2-${ entryName }`,
					map: [
						{ request: 'react', handle: 'react' },
						{ request: 'react-dom', handle: 'react-dom' },
						{ request: /^@elementor\/(.+)$/, handle: 'elementor-v2-$1' },
						{ request: /^@wordpress\/(.+)$/, handle: 'wp-$1' },
					],
				} ),
			],
		} );

		// Expect.
		expect.assertions( 6 );

		// Act.
		compiler.run( ( err, stats ) => {
			// Assert.
			expect( err ).toBe( null );
			expect( stats?.hasErrors() ).toBe( false );
			expect( stats?.hasWarnings() ).toBe( false );

			const files = [ 'app.asset.php', 'extension.asset.php', 'util.asset.php' ];

			files.forEach( ( fileName ) => {
				const fileContent = fs.readFileSync( path.resolve( `/dist/${ fileName }` ), { encoding: 'utf8' } );

				expect( fileContent ).toMatchSnapshot( fileName );
			} );

			done();
		} );
	} );
} );
