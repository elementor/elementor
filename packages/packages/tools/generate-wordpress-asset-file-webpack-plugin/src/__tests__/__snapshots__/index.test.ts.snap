// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@elementor/generate-wordpress-asset-file-webpack-plugin should generate assets files: app.asset.php 1`] = `
"<?php
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}
/**
 * This file is generated by Webpack, do not edit it directly.
 */
return [
	'handle' => 'elementor-v2-app',
	'deps' => [
		'elementor-v2-editor',
		'react',
		'react-dom',
		'wp-element',
	],
];
"
`;

exports[`@elementor/generate-wordpress-asset-file-webpack-plugin should generate assets files: extension.asset.php 1`] = `
"<?php
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}
/**
 * This file is generated by Webpack, do not edit it directly.
 */
return [
	'handle' => 'elementor-v2-extension',
	'deps' => [
		'elementor-v2-editor',
		'react',
		'react-dom',
		'wp-element',
	],
];
"
`;

exports[`@elementor/generate-wordpress-asset-file-webpack-plugin should generate assets files: util.asset.php 1`] = `
"<?php
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}
/**
 * This file is generated by Webpack, do not edit it directly.
 */
return [
	'handle' => 'elementor-v2-util',
	'deps' => [
		'elementor-v2-editor',
		'react',
		'react-dom',
		'wp-element',
	],
];
"
`;
